yapapi/services.py

```diff
@@ -366,39 +408,61 @@ async def _run_instance(self, ctx: WorkContext):
                 (batch_task, signal_task), return_when=asyncio.FIRST_COMPLETED
             )
 
+            def change_state(stop=False, exc_info=(None, None, None)):
```

Comment:
If we stick with the current approach I'd rename this to `advance_state`, see my comment above on `_change_state`.